File: piper/extensions/chrome/src/engine/statements/lists_getIndex.js
import { stringify } from "../util"


export default async function(statement, onError) {
    try {
        let pointer = statement.pointer
        let list = await this.executeStatement(stringify(pointer.list))
        let index = await this.executeStatement(stringify(pointer.index))
        index = parseInt(index)
        let mode = stringify(pointer.mode)
        let where = stringify(pointer.where)

        if (where === "FROM_START") {
            if (mode === "GET") return list[index - 1]
            else if (mode === "GET_REMOVE") return list.splice(index - 1, 1)[0]
            else if (mode === "REMOVE") list.splice(index - 1, 1)
        } else if (where === "FROM_END") {
            if (mode === "GET") return list[list.length - index]
            else if (mode === "GET_REMOVE") return list.splice(list.length - index, 1)[0]
            else if (mode === "REMOVE") list.splice(list.length - index, 1)
        } else if (where === "FIRST") {
            if (mode === "GET") return list[0]
            else if (mode === "GET_REMOVE") return list.splice(0, 1)[0]
            else if (mode === "REMOVE") list.splice(0, 1)
        } else if (where === "LAST") {
            if (mode === "GET") return list[list.length - 1]
            else if (mode === "GET_REMOVE") return list.splice(list.length - 1, 1)[0]
            else if (mode === "REMOVE") list.splice(list.length - 1, 1)
        } else if (where === "RANDOM") {
            let rand = Math.floor(Math.random() * list.length)
            if (mode === "GET") return list[rand]
            else if (mode === "GET_REMOVE") return list.splice(rand, 1)[0]
            else if (mode === "REMOVE") list.splice(rand, 1)
        }
    } catch (e) {
        onError(e)
    }

    return null
}

File: piper/android/app/src/main/java/io/gopiper/piper/engine/statements/lists_getIndex.kt
package io.gopiper.piper.engine.statements

import android.util.Log
import io.gopiper.piper.engine.Piper
import org.json.JSONObject
import java.lang.Exception

suspend fun Piper.executeListsGetIndex(statement: JSONObject, onError: (e: Exception) -> Unit): Any? {
    try {
        Log.d("hello", "execute_lists_getIndex")
        val pointer = statement.getJSONObject("pointer")
        val list = executeStatement(pointer["list"].toString()) as ArrayList<Any?>
        val index = executeStatement(pointer["index"].toString()).toString().toInt()
        val mode = pointer["mode"].toString()

        return when (pointer["where"].toString()) {
            "FROM_START" -> when (mode) {
                "GET" -> list[index - 1]
                "REMOVE", "GET_REMOVE" -> list.removeAt(index - 1)
                else -> null
            }
            "FROM_END" -> when (mode) {
                "GET" -> list[list.size - index]
                "REMOVE", "GET_REMOVE" -> list.removeAt(list.size - index)
                else -> null
            }
            "FIRST" -> when (mode) {
                "GET" -> list[0]
                "REMOVE", "GET_REMOVE" -> list.removeAt(0)
                else -> null
            }
            "LAST" -> when (mode) {
                "GET" -> list.last()
                "REMOVE", "GET_REMOVE" -> list.removeAt(list.lastIndex)
                else -> null
            }
            "RANDOM" -> when (mode) {
                "GET" -> list.random()
                "REMOVE", "GET_REMOVE" -> list.removeAt((0 until list.size).random())
                else -> null
            }
            else -> null
        }
    } catch (e: Exception) {
        onError(e)
        return null
    }
}